src/main/java/org/jenkinsci/remoting/nio/NioChannelHub.java

```diff
@@ -508,11 +497,52 @@ protected CommandTransport makeTransport(InputStream is, OutputStream os, Mode m
         };
     }
 
-    private void scheduleSelectorTask(Callable<Void, IOException> task) {
-        selectorTasks.add(task);
+    // TODO: This logic should use Executor service
+    private void scheduleSelectorTask(java.util.concurrent.Callable<Void> task) {
+        selectorTasks.add(new CallableRemotingWrapper(task));
         selector.wakeup();
     }
 
+    /**
+     * Provides a wrapper for submitting {@link java.util.concurrent.Callable}s over Remoting execution queues.
+     *
+     * @deprecated It is just a hack, which schedules non-serializable tasks over the Remoting Task queue.
+     *             There is no sane reason to reuse this wrapper class anywhere.
+     */
+    @Deprecated
+    private static final class CallableRemotingWrapper implements Callable<Void, IOException> {
+        private static final long serialVersionUID = -7331104479109353930L;
+        final transient java.util.concurrent.Callable<Void> task;
+
+        CallableRemotingWrapper(@Nonnull java.util.concurrent.Callable<Void> task) {
+            this.task = task;
+        }
+
+        @Override
+        public Void call() throws IOException {
+            if (task == null) {
```

Comment:
I personally thing you'd be much better with `assert task != null : "Serialization banned"`

Reply:
I prefer to fail the production code as well. Just in case...